packages/webpack-configs/src/transformers/moduleRules.ts

```diff
@@ -0,0 +1,228 @@
+    matcher.info = {
+        type: matchAssetModuleType.name,
+        value: type
+    };
+
+    return matcher;
+}
+
+export function matchTest(test: string | RegExp): ModuleRuleMatcher {
+    const matcher = (moduleRule: RuleSetRule | RuleSetUseItem) => {
+        if (typeof moduleRule !== "string" && "test" in moduleRule) {
+            if (typeof moduleRule.test === "object" && typeof test === "object") {
+                // Assuming it's regular expressions
+                return moduleRule.test.toString() === test.toString();
+            }
+
+            return moduleRule.test === test;
+        }
+
+        return false;
+    };
+
+    // Add contextual information about the matcher for debugging.
+    matcher.info = {
+        type: matchTest.name,
+        value: test.toString()
+    };
+
+    return matcher;
+}
+
+export interface ModuleRuleMatch {
+    moduleRule: RuleSetRule | RuleSetUseItem;
+    index: number;
+    parent: RuleSetRule[] | RuleSetUseItem[];
+}
+
+function toMatch(moduleRule: RuleSetRule | RuleSetUseItem, index: number, parent: RuleSetRule[] | RuleSetUseItem[]) {
+    return {
+        moduleRule,
+        index,
+        parent
+    };
+}
+
+function isRuleSetRule(value: RuleSetRule | RuleSetUseItem): value is RuleSetRule {
+    return (value as RuleSetRule).use !== undefined || (value as RuleSetRule).oneOf !== undefined;
+}
+
+function findModuleRulesRecursively(moduleRules: RuleSetRule[] | RuleSetUseItem[], matcher: ModuleRuleMatcher, parent: RuleSetRule[] | RuleSetUseItem[], matches: ModuleRuleMatch[]) {
+    moduleRules.forEach((x, index, array) => {
+        if (x) {
+            if (matcher(x, index, array)) {
+                matches.push(toMatch(x, index, parent));
+            } else {
+                if (isRuleSetRule(x)) {
+                    if (x.use) {
+                        findModuleRulesRecursively(x.use as RuleSetUseItem[], matcher, x.use as RuleSetUseItem[], matches);
+                    } else if (x.oneOf) {
+                        findModuleRulesRecursively(x.oneOf, matcher, x.oneOf, matches);
+                    }
+                }
+            }
+        }
+    });
+}
+
+export function findModuleRule(config: Configuration, matcher: ModuleRuleMatcher) {
+    const moduleRules = config.module?.rules;
+
+    if (!moduleRules) {
+        return;
+    }
+
+    const matches: ModuleRuleMatch[] = [];
+
+    findModuleRulesRecursively(moduleRules as RuleSetRule[], matcher, moduleRules as RuleSetRule[], matches);
+
+    if (matches.length > 1) {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        const matcherInfo = matcher.info;
+
+        throw new Error(`[webpack-configs] Found more than 1 matching module rule.\n[webpack-configs] Matcher: "${JSON.stringify(matcherInfo)}"\n[webpack-configs] Matches: "${JSON.stringify(matches.map(x => x.moduleRule))}"`);
+    }
+
+    return matches[0];
```

Comment:
matches[0] but the array might be empty?

Reply:
Yeah, it will return `undefined`, which is the expected output if there's no match.

Reply:
didn't know that was the behavior, interesting thanks